src/Components/InfiniteScroll.tsx

```diff
@@ -0,0 +1,126 @@
+  modelName: string;
+  queryManager: IQueryManager<T>;
+  modelComponent: React.ComponentType<{ model: T } & any>;
+  modelComponentProps?: any;
+}
+
+const DisplayPreviousPage = compose<IDisplayPreviousPageProps<any>, IDisplayPreviousPageProps<any>>(
+  setDisplayName("DisplayPreviousPage"),
+  pure,
+)(
+  (({ queryManager, modelName, modelComponent: ModelComponent, modelComponentProps }) => {
+    return queryManager.hasPreviousPage() && (
+      <Query key={`page-${queryManager.response.previousPage}`} modelName={modelName} query={queryManager.getPreviousPage()}>
+        {({ query }) => (
+          <>
+            <DisplayPreviousPage queryManager={query} modelComponent={ModelComponent} modelComponentProps={modelComponentProps} modelName={modelName} />
+            {query.items.map(model => (
+              <ModelComponent key={model.id} model={model} {...modelComponentProps} />
+            ))}
+          </>
+        )}
+      </Query>
+    );
+  }),
+);
+
+export const InfiniteScroll = compose<IInfiniteScrollInternalProps<any>, IInfiniteScrollProps<any>>(
+  setDisplayName("InfiniteScroll"),
+  defaultProps({
+    debounceTime: 200,
+  }),
+  withState<IInfiniteScrollProps<any>, IQueryBuilder, "query", "updateQuery">("query", "updateQuery", ({ query, modelName }) => (
+    query instanceof QueryBuilder
+      ? query
+      : new QueryBuilder(modelName, query as IQueryParams)
+  )),
+  withModelQuery(),
+  withStateHandlers<{ lastScrollTop: number }, { handleScroll }, IInfiniteScrollInternalProps<any>>(
+    { lastScrollTop: 0 },
+    {
+      handleScroll: ({ lastScrollTop }, { query, updateQuery }) => (clientHeight: number, scrollHeight: number, currentScrollTop: number) => {
+        const scrollingDown = currentScrollTop > lastScrollTop;
+        const scrollBottom = scrollHeight - currentScrollTop - clientHeight;
+
+        if (scrollingDown && scrollBottom < (clientHeight / 2) && query.hasNextPage()) {
+          updateQuery(query.getNextPage());
+        }
+
+        return { lastScrollTop: currentScrollTop };
+      },
+    },
+  ),
+  withProps(({ handleScroll, debounceTime }) => ({
+    handleScrollDebounced: debounce(handleScroll, debounceTime),
+  })),
+  withProps(({ handleScrollDebounced }) => ({
+    handleScrollPersistingEvent: (event: any) => {
+      const clientHeight = event.target.clientHeight;
+      const scrollHeight = event.target.scrollHeight;
+      const scrollTop = event.target.scrollTop;
+
+      handleScrollDebounced(clientHeight, scrollHeight, scrollTop);
+    },
+  })),
+  lifecycle<{ handleScrollDebounced }, {}>({
+    componentWillUnmount() {
+      this.props.handleScrollDebounced.cancel();
+    },
+  }),
+)(({
+  containerComponent: ContainerComponent,
+  containerProps,
+  modelComponent: ModelComponent,
+  modelComponentProps,
+  query: queryManager,
+  modelName,
+  handleScrollPersistingEvent,
+}) => (
+    <ContainerComponent {...containerProps} onScroll={handleScrollPersistingEvent}>
+      <>
```

Comment:
This React Fragment should be removed. The Material-UI Select component will have issues because of it.

Reply:
removed